packages/core/src/rules/common/no-required-schema-properties-undefined.ts

```diff
@@ -0,0 +1,52 @@
+import { Oas2Rule, Oas3Rule } from '../../visitors';
+import { Oas3Schema, Oas3_1Schema } from '../../typings/openapi';
+import { Oas2Schema } from 'core/src/typings/swagger';
+import { UserContext } from 'core/src/walk';
+import { isRef } from '../../ref-utils';
+
+export const NoRequiredSchemaPropertiesUndefined: Oas3Rule | Oas2Rule = () => {
+  return {
+    Schema: {
+      enter(
+        schema: Oas3Schema | Oas3_1Schema | Oas2Schema,
+        { location, report, resolve }: UserContext
+      ) {
+        if (!schema.required) return;
+        const visitedSchemas: Set<Oas3Schema | Oas3_1Schema | Oas2Schema> = new Set();
+
+        const elevateProperties = (
+          schema: Oas3Schema | Oas3_1Schema | Oas2Schema
+        ): Record<string, Oas3Schema | Oas3_1Schema | Oas2Schema> => {
+          // Check if the schema has been visited before processing it
+          if (visitedSchemas.has(schema)) {
+            return {};
+          }
+          visitedSchemas.add(schema);
+
+          if (isRef(schema)) {
+            return elevateProperties(
+              resolve(schema).node as Oas3Schema | Oas3_1Schema | Oas2Schema
+            );
+          }
+
+          return Object.assign(
+            {},
+            schema.properties,
+            ...(schema.allOf?.map(elevateProperties) ?? [])
+          );
+        };
+
+        const allProperties = elevateProperties(schema);
```

Comment:
can we add test for nested allOf and recursive allOf?

Reply:
Added two more test cases for nested allOf schema examples